kotlinx-coroutines-core/common/src/flow/operators/Delay.kt

```diff
@@ -32,3 +35,125 @@ public fun <T> Flow<T>.delayEach(timeMillis: Long): Flow<T> = flow {
         emit(value)
     }
 }
+
+/**
+ * Returns a flow that mirrors the original flow, but filters out values
+ * that are followed by the newer values within the given [timeout][timeoutMs].
+ * The latest value is always emitted.
+ * Example:
+ * ```
+ * flow {
+ *   emit(1)
+ *   delay(99)
+ *   emit(2)
+ *   delay(99)
+ *   emit(3)
+ *   delay(1001)
+ *   emit(4)
+ *   delay(1001)
+ *   emit(5)
+ * }.debounce(1000)
+ * ```
+ * will produce `3, 4, 5`.
+ */
+public fun <T> Flow<T>.debounce(timeoutMs: Long): Flow<T> {
+    require(timeoutMs > 0) { "Debounce timeout should be positive" }
+    return flow {
+        coroutineScope {
+            val values = Channel<Any>(Channel.CONFLATED)
+            // Channel is not closed deliberately as there is no close with value
+            val collector = launch {
+                try {
+                    collect { value -> values.send(value ?: NullSurrogate) }
+                } catch (e: Throwable) {
+                    values.close(e) // Workaround for #1130
+                    throw e
+                }
+            }
+
+            var isDone = false
+            var lastValue: Any? = null
+            while (!isDone) {
+                select<Unit> {
+                    values.onReceive {
+                        lastValue = it
+                    }
+
+                    onTimeout(timeoutMs) {
```

Comment:
As is, won't this select statement run every `timeoutMs` for an idle Flow? Wrapping with `if (lastValue != null)` still passes all the unit tests and should avoid extra checking? The same question applies to `Flow<T>.sample` and the `ticker.onReceive` select statement below.

Reply:
I'm taking care of it.